src/lib/storage/zero_suppression/simd_bp128/simd_bp128_decoder.cpp

```diff
@@ -0,0 +1,43 @@
+#include "simd_bp128_decoder.hpp"
+
+#include "simd_bp128_vector.hpp"
+
+namespace opossum {
+
+SimdBp128Decoder::SimdBp128Decoder(const SimdBp128Vector& vector)
+    : _data{&vector._data},
+      _size{vector._size},
+      _cached_meta_info_offset{0u},
+      _cached_meta_block_first_index{std::numeric_limits<size_t>::max()},
+      _cached_block_first_index{std::numeric_limits<size_t>::max()},
+      _cached_block{std::make_unique<std::array<uint32_t, Packing::block_size>>()} {}
+
+SimdBp128Decoder::SimdBp128Decoder(const SimdBp128Decoder& other)
+    : _data{other._data},
+      _size{other._size},
+      _cached_meta_info_offset{other._cached_meta_info_offset},
+      _cached_meta_block_first_index{other._cached_meta_block_first_index},
+      _cached_meta_info{other._cached_meta_info},
+      _cached_block_first_index{other._cached_block_first_index},
+      _cached_block{std::make_unique<std::array<uint32_t, Packing::block_size>>(*other._cached_block)} {}
+
+void SimdBp128Decoder::_read_meta_info(size_t meta_info_offset) {
+  Packing::read_meta_info(_data->data() + meta_info_offset, _cached_meta_info.data());
+}
+
+void SimdBp128Decoder::_unpack_block(uint8_t block_index) {
+  static const auto meta_info_data_size = 1u;  // One 128 bit block
+  const auto relative_data_offset =
+      meta_info_data_size + std::accumulate(_cached_meta_info.begin(), _cached_meta_info.begin() + block_index, 0u);
+  const auto data_offset = _cached_meta_info_offset + relative_data_offset;
+
+  const auto in = _data->data() + data_offset;
+  auto out = _cached_block->data();
+  const auto bit_size = _cached_meta_info[block_index];
+
+  Packing::unpack_block(in, out, bit_size);
+
+  _cached_block_first_index = _cached_meta_block_first_index + block_index * Packing::block_size;
```

Comment:
?

Reply:
I added explanations to the several member variables. Hope that makes it clearer.